tests/functional/gin.cpp

```diff
@@ -0,0 +1,268 @@
+
+	if (test_support_gdr[dev] == 1) {
+		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET,
+				"Network supports communication using CUDA buffers. Dev: %d", dev);
+		buffer_type = NCCL_PTR_CUDA;
+	} else {
+		/* We aren't currently interested in the non-GDR use case for GIN */
+		NCCL_OFI_WARN("Network does not support communication using CUDA buffers. Dev: %d", dev);
+		return 1;
+	}
+
+	void *listenComm = nullptr;
+	OFINCCLCHECK(extGin->listen(ginCtx, dev, handles[rank], &listenComm));
+	assert(listenComm);
+
+	/* Gather handles from all ranks */
+	MPI_Allgather(MPI_IN_PLACE, 0, MPI_DATATYPE_NULL, handles.data(), NCCL_NET_HANDLE_MAXSIZE, MPI_CHAR, MPI_COMM_WORLD);
+
+	/* Prepare handles void** array */
+	for (int i = 0; i < nranks; ++i) {
+		handles_ptrs[i] = &(handles[i]);
+	}
+
+	void *collComm = nullptr;
+	OFINCCLCHECK(extGin->connect(ginCtx, handles_ptrs.data(), nranks, rank, listenComm,
+				     &collComm));
+	assert(collComm != nullptr);
+
+	void *buff = nullptr;
+	OFINCCLCHECK(allocate_buff(&buff, SEND_SIZE, buffer_type));
+	OFINCCLCHECK(initialize_buff(buff, SEND_SIZE, buffer_type));
+
+	/* Collective memory registration */
+	void *mhandle = nullptr, *gin_mhandle = nullptr;
+	constexpr uint64_t mrFlags = 0; /* TODO FORCE_SO */
+	OFINCCLCHECK(extGin->regMrSym(collComm, buff, SEND_SIZE, buffer_type, mrFlags,
+				      &mhandle, &gin_mhandle));
+	assert(mhandle != nullptr && gin_mhandle != nullptr);
+
+	void *signal_buf = nullptr;
+	OFINCCLCHECK(allocate_buff(&signal_buf, sizeof(uint64_t), buffer_type));
+	OFINCCLCHECK(initialize_buff(buff, SEND_SIZE, buffer_type, 0));
+	CUDACHECK(cudaStreamSynchronize(cudaStreamDefault));
+
+	void *signalMhandle = nullptr, *gin_signalMhandle = nullptr;
+	OFINCCLCHECK(extGin->regMrSym(collComm, signal_buf, sizeof(uint64_t), buffer_type, mrFlags,
+				      &signalMhandle, &gin_signalMhandle));
+	assert(signalMhandle != nullptr && gin_signalMhandle != nullptr);
+
+	const int send_val = 42; /* because */
+	const int NUM_REQS_PER_PEER = 64;
+	assert(NUM_REQS_PER_PEER < NCCL_OFI_MAX_REQUESTS);
+
+	if (rank == 0) {
+		CUDACHECK(cudaMemset(buff, send_val, SEND_SIZE));
+		CUDACHECK(cudaStreamSynchronize(cudaStreamDefault));
+
+		std::deque<void *> request_deque;
+
+		for (int dst_rank = 1; dst_rank < nranks; ++dst_rank) {
+
+			/* iputSignal API */
+
+			for (int i = 0; i < NUM_REQS_PER_PEER; ++i)
+			{
+				void *request = nullptr;
+				OFINCCLCHECK(extGin->iputSignal(collComm, 0, mhandle, SEND_SIZE, 0, mhandle,
+					dst_rank, 0, gin_signalMhandle, 1, NCCL_NET_SIGNAL_OP_INC, &request));
+				assert(request != nullptr);
+				request_deque.push_back(request);
+			}
+		}
+
+		/* Wait for remaining requests */
+		while (!request_deque.empty()) {
+			OFINCCLCHECK(poll_request_completion(extGin, request_deque, collComm));
+		}
+	} else {
+		uint64_t signal_h = 0;
+		while (signal_h != NUM_REQS_PER_PEER) {
+			OFINCCLCHECK(extGin->ginProgress(collComm));
```

Comment:
Rank 0 calls `ginProgress` by calling `poll_request_completion` as well as all the other ranks. Are both required for the test to correctly complete?

For example, if the initiator side call to `ginProgress` is not required, suggest adding a comment.

If both are needed, even then a comment would be useful.